### ISSUE TITLE
mojo_test with enable-multiprocess does not exit gracefully when app.close() is called.

### ISSUE THREAD

**Issue by aghassemi** [domokit/mojo#487]
When running `mojo_test` with `--enable-multiprocess`, doing a `await app.close()` cause the whole process to exit with error and no code following  `await app.close()` would run.

This does not happen without `--enable-multiprocess`. 

This is important since normally tests want to do a `reportLeakedHandles()` after `app.close()` but with 
 `--enable-multiprocess` that is not possible currently.

0:00 +0: terminate shell connection
[ERROR:child_main.cc(245)] Connection error to the shell
[ERROR:child_main.cc(245)] Connection error to the shell
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483689, remote ID 43)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483651, remote ID 5)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483671, remote ID 25)
[WARNING:channel.cc(335)] RawChannel write error
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483672, remote ID 26)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483669, remote ID 23)
[ERROR:child_main.cc(245)] Connection error to the shell
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483690, remote ID 44)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483691, remote ID 45)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483653, remote ID 7)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483661, remote ID 15)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483652, remote ID 6)
[WARNING:channel.cc(571)] Failed to send message to ack remove remote endpoint (local ID 2147483663, remote ID 17)
[ERROR:child_main.cc(158)] Disconnected from master
[ERROR:child_main.cc(158)] Disconnected from master
`

**Comment by zanderso** [comment]
I think we probably need to add a special cases to the Dart Mojo libraries for handles to the shell. We shouldn't be closing them from Dart code, but rather from DartApp::OnAppLoaded after DartController::RunDartScript comes back.

**Comment by jamesr** [comment]
Could the Dart application framework do the wrapping of the shell handle itself and exposed the wrapped type to application code?

**Comment by zanderso** [comment]
We pass the raw integer handle to the Dart application's main() function, where it is then typically wrapped by the types from the Dart application framework. We can't do the wrapping before invoking the Dart main() function since the wrapper types aren't guaranteed to exist. They are in package:mojo, which an app isn't obligated to import. What I think we should do is get Dart's application library to tell the wrapper types that the shell handle is special, and not to worry about closing it, and then take care of closing it in DartApp in C++ after we come back from Dart code.

**Comment by jamesr** [comment]
I don't think we should specialize the wrapper types.  We can make package:mojo do the right thing in main and if developers want to do something different for their main we can document the requirements on dealing with that handle.

**Comment by zanderso** [comment]
I'm not sure what you mean, but after looking at the code some more, what I'm thinking is that we change this _stub.close():

https://github.com/domokit/mojo/blob/master/mojo/public/dart/mojo/lib/src/application.dart#L49

To be a new call into Stub that removes the handle from our leak checking, but which doesn't do the low-level close() on it. We might also need to do something else to tell the embedder not to try to close the handle on an unhandled exception, but since I'm currently cleaning that code up, I need to think about it more.

**Comment by jamesr** [comment]
Ah, we do wrap it.  I thought you were proposing having the code generator produce different logic in application.mojom.dart
